### ISSUE TITLE
Glue

### ISSUE THREAD

**Issue by yakir12** [JuliaDynamics/DynamicalBilliards.jl#173]
Could be realted to #153 
Consider this example: 
```julia
using DynamicalBilliards, PyPlot
bdr =  billiard_rectangle(6,4)
bd = Billiard(Semicircle([2,2], 1, [0,1], "mirror1"), Semicircle([4,2], 1, [0,-1], "mirror2"), bdr.obstacles...)
plot(bd)
for y in range(1.5, stop = 2.5, length = 11)
    p = Particle(4.0, y, pi)
    xt, yt, vxt, vyt, t = timeseries!(p, bd, 1)
    plot(xt, yt)
end
```
Which results in this:
![screenshot_2018-11-05_15 19 50](https://user-images.githubusercontent.com/3281957/48003309-69770680-e10e-11e8-9be5-ace5706a118a.png)

I find the middle ray (the one where `y==2`) to be problematic. It seems as if it missed both semicircles. I can understand why this happens, but I think it's problematic since the two semicircles are "supposed" to touch. This does not seem to happen with standard billiards, where a particle would escape out (which it doesn't).

The intended use-case here is to construct more complex, piece-wise, obstacles that are made of glued obstacles...

**Comment by yakir12** [comment]
Hmm, I guess this can be solved by defining a new obstacle that just "does this".

**Comment by Datseris** [comment]
TLDR;

If you change your code from `p = Particle(4, y, pi)` to `p = Particle(4.0, y - 1e-15, pi)` the "problem" is resolved.

---

Real answer:

There is no problem stemming form `DynamicalBilliards` in your example. There are two issues that you hit simultaneously here:
1. Finite precision, which all billiards of the world suffer from it and somehow have to get around it with one way or the other. 
2. Mathematical problem: collision with the end of an obstacle is an **undefined** process. Rigorously speaking, a particle colliding with exactly the end of an obstacle (or the point where two obstacles "combine/glue") cannot be evolved further. Its evolution terminates there.

Now, depending on the package, different things may happen. In `DynamicalBilliards` you can never collide with the true end of your obstacle. You either miss it completely or you collide normally with the next floating point point before the end. In your above example you missed it completely.

---
Now on to the glued obstacles.

> The intended use-case here is to construct more complex, piece-wise, obstacles that are made of glued obstacles...

Your code already does this! There is no "gluing" process in `DynamicalBilliards`, you simply put the obstacles next to each other. I have never tested what would happen in the Sinai billiard if an orbit with **exactly** π/4 angle of incidence went **exactly** into the corner of the billiard. It may do a normal collision, it may slip through.

If you truly, TRULY care about the single trajectory that can do this in your specific case, just put the second semicircle epsilon bit down. 

---

p.s.: The `Semicircle` obstacle cannot do what you want to do. It doesn't allow for raysplitting and as a result it only supports particles that are inside the Semicircle. You would have to create a semicircle that also supports raysplitting to have a proper code that manifests your above example.


I am closing this, as it is not an issue with `DynamicalBilliards` but with the mathematical definition of a billiard. For more discussion on "gluing" obstacles let's move this to #153

**Comment by Datseris** [comment]
> Hmm, I guess this can be solved by defining a new obstacle that just "does this".

Yes that is true, you can do this, if you _really_ care about singular orbits. It is not possible to handle them "generally" for every obstacle and billiard combination though.
